<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.27">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="Predict kinases given a substrate sequence">

<title>KATLAS – python-katlas</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdn.jsdelivr.net/npm/jquery@3.5.1/dist/jquery.min.js" integrity="sha384-ZvpUoO/+PpLXR1lu4jmpXWu80pZlYUAfxl5NsBMWOEPSjUn/6Z/hRTt8+pR6L4N2" crossorigin="anonymous"></script><script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-ed96de9b727972fe78a7b5d16c58bf87.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-45a6b1ee3682cc1dbb0c772b170fe9b7.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdn.jsdelivr.net/npm/requirejs@2.3.6/require.min.js" integrity="sha384-c9c+LnTbwQ3aujuU7ULEPVvgLs+Fn6fJUvIGTsuu1ZcCf11fiEubah0ttpca4ntM sha384-6V1/AdqZRWk1KAlWbKBlGhN7VG4iE/yAZcO6NZPMF8od0vukrvr0tg4qY6NSrItx" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="KATLAS – python-katlas">
<meta property="og:description" content="Predict kinases given a substrate sequence">
<meta property="og:site_name" content="python-katlas">
<meta name="twitter:title" content="KATLAS – python-katlas">
<meta name="twitter:description" content="Predict kinases given a substrate sequence">
<meta name="twitter:card" content="summary">
</head>

<body class="nav-sidebar floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">python-katlas</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/sky1ove/katlas"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./index.html">Get Started</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Get Started</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Modules</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./utils.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Utils</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pssm_core.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">pssm.core</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pssm_plot.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">pssm.plot</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pssm_lo.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">pssm.lo</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pssm_pspa.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">pssm.pspa</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pssm_compare.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">pssm.compare</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./scoring.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Scoring</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Hierarchical clustering</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./hierarchical.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Hierarchical clustering</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./feature.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Feature</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./plot.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Plot</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pathway.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Pathway analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ml.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Train ML</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./dnn.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Train DL</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#reproduce-datasets-figures" id="toc-reproduce-datasets-figures" class="nav-link active" data-scroll-target="#reproduce-datasets-figures">Reproduce datasets &amp; figures</a></li>
  <li><a href="#web-applications" id="toc-web-applications" class="nav-link" data-scroll-target="#web-applications">Web applications</a></li>
  <li><a href="#install" id="toc-install" class="nav-link" data-scroll-target="#install">Install</a></li>
  <li><a href="#import" id="toc-import" class="nav-link" data-scroll-target="#import">Import</a></li>
  <li><a href="#quick-start" id="toc-quick-start" class="nav-link" data-scroll-target="#quick-start">Quick start</a>
  <ul class="collapse">
  <li><a href="#quick-start-1" id="toc-quick-start-1" class="nav-link" data-scroll-target="#quick-start-1">Quick start</a>
  <ul class="collapse">
  <li><a href="#site-scoring" id="toc-site-scoring" class="nav-link" data-scroll-target="#site-scoring">Site scoring</a></li>
  <li><a href="#site-scoring-in-a-df" id="toc-site-scoring-in-a-df" class="nav-link" data-scroll-target="#site-scoring-in-a-df">Site scoring in a df</a></li>
  </ul></li>
  <li><a href="#dataset" id="toc-dataset" class="nav-link" data-scroll-target="#dataset">Dataset</a>
  <ul class="collapse">
  <li><a href="#cptac-pan-cancer-phosphoproteomics" id="toc-cptac-pan-cancer-phosphoproteomics" class="nav-link" data-scroll-target="#cptac-pan-cancer-phosphoproteomics">CPTAC pan-cancer phosphoproteomics</a></li>
  <li><a href="#ochoa-et-al.-human-phosphoproteome" id="toc-ochoa-et-al.-human-phosphoproteome" class="nav-link" data-scroll-target="#ochoa-et-al.-human-phosphoproteome">Ochoa et al.&nbsp;human phosphoproteome</a></li>
  <li><a href="#phosphositeplus-human-phosphorylation-site" id="toc-phosphositeplus-human-phosphorylation-site" class="nav-link" data-scroll-target="#phosphositeplus-human-phosphorylation-site">PhosphoSitePlus human phosphorylation site</a></li>
  <li><a href="#unique-sites-of-combined-ochoa-phosphositeplus" id="toc-unique-sites-of-combined-ochoa-phosphositeplus" class="nav-link" data-scroll-target="#unique-sites-of-combined-ochoa-phosphositeplus">Unique sites of combined Ochoa &amp; PhosphoSitePlus</a></li>
  </ul></li>
  <li><a href="#phosphorylation-site-sequence-example" id="toc-phosphorylation-site-sequence-example" class="nav-link" data-scroll-target="#phosphorylation-site-sequence-example">Phosphorylation site sequence example</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/sky1ove/katlas/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">KATLAS</h1>
</div>

<div>
  <div class="description">
    Predict kinases given a substrate sequence
  </div>
</div>


<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<p><img alt="Katlas logo" width="600" caption="Katlas logo" src="https://github.com/sky1ove/katlas/raw/main/logo.png" id="logo"></p>
<p>KATLAS is a repository containing python tools to predict kinases given a substrate sequence. It also contains datasets of kinase substrate specificities and human phosphoproteomics.</p>
<p><strong><em>References</em></strong>: Please cite the appropriate papers if KATLAS is helpful to your research.</p>
<ul>
<li><p>KATLAS was described in the paper [Computational Decoding of Human Kinome Substrate Specificities and Functions]</p></li>
<li><p>The positional scanning peptide array (PSPA) data is from paper <a href="https://www.nature.com/articles/s41586-022-05575-3">An atlas of substrate specificities for the human serine/threonine kinome</a> and paper <a href="https://www.nature.com/articles/s41586-024-07407-y">The intrinsic substrate specificity of the human tyrosine kinome</a></p></li>
<li><p>The kinase substrate datasets used for generating PSSMs are derived from <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3245126/">PhosphoSitePlus</a> and paper <a href="https://www.nature.com/articles/s41598-019-46385-4">Large-scale Discovery of Substrates of the Human Kinome</a></p></li>
<li><p>Phosphorylation sites are acquired from <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3245126/">PhosphoSitePlus</a>, paper <a href="https://www.nature.com/articles/s41587-019-0344-3">The functional landscape of the human phosphoproteome</a>, and <a href="https://pdc.cancer.gov/pdc/cptac-pancancer">CPTAC</a> / <a href="https://academic.oup.com/nar/article/46/D1/D956/4607804">LinkedOmics</a></p></li>
</ul>
<section id="reproduce-datasets-figures" class="level2">
<h2 class="anchored" data-anchor-id="reproduce-datasets-figures">Reproduce datasets &amp; figures</h2>
<p>Follow the instructions in katlas_raw: https://github.com/sky1ove/katlas_raw</p>
</section>
<section id="web-applications" class="level2">
<h2 class="anchored" data-anchor-id="web-applications">Web applications</h2>
<p>Users can now run the analysis directly on the web without needing to code.</p>
<p>Check out our latest web platform: <a href="https://kinase-atlas.com/">kinase-atlas.com</a></p>
</section>
<section id="install" class="level2">
<h2 class="anchored" data-anchor-id="install">Install</h2>
<p>UV:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">uv</span> add <span class="at">-U</span> python-katlas</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>pip:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install <span class="at">-U</span> python-katlas</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>If using machine-learning related modules, need to install development verison: <code>pip install -U "python-katlas[dev]"</code></p>
</section>
<section id="import" class="level2">
<h2 class="anchored" data-anchor-id="import">Import</h2>
<div id="5392550f" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> katlas.common <span class="im">import</span> <span class="op">*</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="quick-start" class="level1">
<h1>Quick start</h1>
<p>We provide two methods to calculate substrate sequence:</p>
<ul>
<li>Computational Data-Driven Method (CDDM)</li>
<li>Positional Scanning Peptide Array (PSPA)</li>
</ul>
<p>We consider the input in two formats:</p>
<ul>
<li>a single input string (phosphorylation site)</li>
<li>a csv/dataframe that contains a column of phosphorylation sites</li>
</ul>
<p>For input sequences, we also consider it in two conditions:</p>
<ul>
<li>all capital</li>
<li>contains lower cases indicating phosphorylation status</li>
</ul>
<section id="quick-start-1" class="level2">
<h2 class="anchored" data-anchor-id="quick-start-1">Quick start</h2>
<section id="site-scoring" class="level3">
<h3 class="anchored" data-anchor-id="site-scoring">Site scoring</h3>
<p>CDDM, all capital</p>
<div id="5f165c16" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>predict_kinase(<span class="st">'AAAAAAASGAGSDN'</span>,<span class="op">**</span>Params(<span class="st">"CDDM_upper"</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>considering string: ['-7A', '-6A', '-5A', '-4A', '-3A', '-2A', '-1A', '0S', '1G', '2A', '3G', '4S', '5D', '6N']</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>GCN2      4.556
MPSK1     4.425
MEKK2     4.253
WNK3      4.213
WNK1      4.064
          ...  
PDK1    -25.077
PDHK3   -25.346
CLK2    -27.251
ROR2    -27.582
DDR1    -53.581
Length: 328, dtype: float64</code></pre>
</div>
</div>
<p>CDDM, with lower case indicating phosphorylation status</p>
<div id="97566af5" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>predict_kinase(<span class="st">'AAAAAAAsGGAGsDN'</span>,<span class="op">**</span>Params(<span class="st">"CDDM"</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>considering string: ['-7A', '-6A', '-5A', '-4A', '-3A', '-2A', '-1A', '0s', '1G', '2G', '3A', '4G', '5s', '6D', '7N']</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>ROR1       8.355
WNK1       4.907
WNK2       4.782
ERK5       4.466
RIPK2      4.045
           ...  
DDR1     -29.393
TNNI3K   -29.884
CHAK1    -31.775
VRK1     -45.287
BRAF     -49.403
Length: 328, dtype: float64</code></pre>
</div>
</div>
<p>PSPA, with lower case indicating phosphorylation status</p>
<div id="21ed7384" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>predict_kinase(<span class="st">'AEEKEyHsEGG'</span>,<span class="op">**</span>Params(<span class="st">"PSPA"</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>considering string: ['-5A', '-4E', '-3E', '-2K', '-1E', '0y', '1H', '2s', '3E', '4G', '5G']</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>kinase
EGFR          4.013
FGFR4         3.568
ZAP70         3.412
CSK           3.241
SYK           3.209
              ...  
JAK1         -3.837
DDR2         -4.421
TNK2         -4.534
TNNI3K_TYR   -4.651
TNK1         -5.320
Length: 93, dtype: float64</code></pre>
</div>
</div>
<p>To replicate the results from The Kinase Library (PSPA)</p>
<p>Check this link: <a href="https://kinase-library.mit.edu/site?s=AEEKEy*HSEGG&amp;pp=false&amp;scp=true">The Kinase Library</a>, and use log2(score) to rank, it shows same results with the below (with slight differences due to rounding).</p>
<div id="e79d88a0" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>out <span class="op">=</span> predict_kinase(<span class="st">'AEEKEyHSEGG'</span>,<span class="op">**</span>Params(<span class="st">"PSPA"</span>))</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>out</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>considering string: ['-5A', '-4E', '-3E', '-2K', '-1E', '0y', '1H', '2S', '3E', '4G', '5G']</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>kinase
EGFR     3.181
FGFR4    2.390
CSK      2.308
ZAP70    2.068
SYK      1.998
         ...  
EPHA1   -3.501
FES     -3.699
TNK1    -4.269
TNK2    -4.577
DDR2    -4.920
Length: 93, dtype: float64</code></pre>
</div>
</div>
<ul>
<li>So far <a href="https://kinase-library.phosphosite.org">The kinase Library</a> considers all <strong><em>tyr sequences</em></strong> in capital regardless of whether or not they contain lower cases, which is a small bug and should be fixed soon.</li>
<li>Kinase with “_TYR” indicates it is a dual specificity kinase tested in PSPA tyrosine setting, which has not been included in kinase-library yet.</li>
</ul>
<p>We can also calculate the percentile score using a referenced score sheet.</p>
<div id="0b74a8c7" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Percentile reference sheet</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>y_pct <span class="op">=</span> Data.get_pspa_tyr_pct()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="29e63f9f" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>get_pct(<span class="st">'AEEKEyHSEGG'</span>,pct_ref <span class="op">=</span> y_pct,<span class="op">**</span>Params(<span class="st">"PSPA_y"</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>considering string: ['-5A', '-4E', '-3E', '-2K', '-1E', '0Y', '1H', '2S', '3E', '4G', '5G']</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">log2(score)</th>
<th data-quarto-table-cell-role="th">percentile</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">EGFR</th>
<td>3.181</td>
<td>96.787423</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">FGFR4</th>
<td>2.390</td>
<td>94.012303</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">CSK</th>
<td>2.308</td>
<td>95.201640</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">ZAP70</th>
<td>2.068</td>
<td>88.380041</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">SYK</th>
<td>1.998</td>
<td>85.522898</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">...</th>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">EPHA1</th>
<td>-3.501</td>
<td>12.139440</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">FES</th>
<td>-3.699</td>
<td>21.216678</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">TNK1</th>
<td>-4.269</td>
<td>5.481887</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">TNK2</th>
<td>-4.577</td>
<td>2.050581</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">DDR2</th>
<td>-4.920</td>
<td>10.403281</td>
</tr>
</tbody>
</table>

<p>93 rows × 2 columns</p>
</div>
</div>
</div>
</section>
<section id="site-scoring-in-a-df" class="level3">
<h3 class="anchored" data-anchor-id="site-scoring-in-a-df">Site scoring in a df</h3>
<p>Load your csv:</p>
<div id="b76b28cf" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># df = pd.read_csv('your_file.csv')</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Or load a demo df</p>
<div id="ab133342" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load a demo df with phosphorylation sites</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> Data.get_ochoa_site().head()</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>df.iloc[:,<span class="op">-</span><span class="dv">2</span>:]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">site_seq</th>
<th data-quarto-table-cell-role="th">gene_site</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">0</th>
<td>VDDEKGDSNDDYDSA</td>
<td>A0A075B6Q4_S24</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>YDSAGLLSDEDCMSV</td>
<td>A0A075B6Q4_S35</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>IADHLFWSEETKSRF</td>
<td>A0A075B6Q4_S57</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">3</th>
<td>KSRFTEYSMTSSVMR</td>
<td>A0A075B6Q4_S68</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">4</th>
<td>FTEYSMTSSVMRRNE</td>
<td>A0A075B6Q4_S71</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>Set the column name and param to calculate</p>
<p>Here we choose param_CDDM_upper, as the sequences in the demo df are all in capital. You can also choose other params.</p>
<div id="895000bd" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>results <span class="op">=</span> predict_kinase_df(df,<span class="st">'site_seq'</span>,<span class="op">**</span>Params(<span class="st">"CDDM_upper"</span>))</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>results</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>input dataframe has a length 5
Preprocessing
Finish preprocessing
Merging reference
Finish merging</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">SRC</th>
<th data-quarto-table-cell-role="th">EPHA3</th>
<th data-quarto-table-cell-role="th">FES</th>
<th data-quarto-table-cell-role="th">NTRK3</th>
<th data-quarto-table-cell-role="th">ALK</th>
<th data-quarto-table-cell-role="th">ABL1</th>
<th data-quarto-table-cell-role="th">FLT3</th>
<th data-quarto-table-cell-role="th">EPHA8</th>
<th data-quarto-table-cell-role="th">EPHB2</th>
<th data-quarto-table-cell-role="th">EPHB1</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">VRK1</th>
<th data-quarto-table-cell-role="th">PKMYT1</th>
<th data-quarto-table-cell-role="th">GRK3</th>
<th data-quarto-table-cell-role="th">CAMK1B</th>
<th data-quarto-table-cell-role="th">CDC7</th>
<th data-quarto-table-cell-role="th">SMMLCK</th>
<th data-quarto-table-cell-role="th">ROR1</th>
<th data-quarto-table-cell-role="th">GAK</th>
<th data-quarto-table-cell-role="th">MAST2</th>
<th data-quarto-table-cell-role="th">BRAF</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">0</th>
<td>-2.440640</td>
<td>-0.818753</td>
<td>-1.663990</td>
<td>-0.738991</td>
<td>-2.047628</td>
<td>-3.602344</td>
<td>-3.200998</td>
<td>-0.935176</td>
<td>-1.388444</td>
<td>-1.859450</td>
<td>...</td>
<td>-17.103237</td>
<td>-113.698143</td>
<td>-16.848783</td>
<td>-41.520172</td>
<td>-41.646187</td>
<td>1.284159</td>
<td>-26.566362</td>
<td>-69.165062</td>
<td>-17.706400</td>
<td>-87.763214</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>-3.838486</td>
<td>-2.735969</td>
<td>-2.533986</td>
<td>-2.150399</td>
<td>-3.792498</td>
<td>-4.725527</td>
<td>-5.711791</td>
<td>-4.534240</td>
<td>-3.148449</td>
<td>-2.511518</td>
<td>...</td>
<td>-67.889053</td>
<td>-68.652641</td>
<td>-45.833855</td>
<td>-64.171600</td>
<td>-39.465572</td>
<td>-65.061722</td>
<td>-109.561707</td>
<td>-85.911224</td>
<td>-60.105064</td>
<td>-63.889122</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>-2.610423</td>
<td>-2.370090</td>
<td>-3.235637</td>
<td>-1.508413</td>
<td>-2.571347</td>
<td>-3.740941</td>
<td>-3.025596</td>
<td>-3.373504</td>
<td>-2.776297</td>
<td>-3.060740</td>
<td>...</td>
<td>-15.798462</td>
<td>-45.905319</td>
<td>-61.440742</td>
<td>-67.695694</td>
<td>-55.047962</td>
<td>-42.135216</td>
<td>-38.501572</td>
<td>-62.624382</td>
<td>-56.119389</td>
<td>-107.060989</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">3</th>
<td>-5.180541</td>
<td>-4.201880</td>
<td>-5.766463</td>
<td>-3.038421</td>
<td>-3.836897</td>
<td>-4.249900</td>
<td>-5.029885</td>
<td>-5.411311</td>
<td>-4.713308</td>
<td>-4.827825</td>
<td>...</td>
<td>-96.978317</td>
<td>-83.419777</td>
<td>-22.559393</td>
<td>-110.611588</td>
<td>-63.283070</td>
<td>-37.240440</td>
<td>-24.497492</td>
<td>-112.878151</td>
<td>-43.538158</td>
<td>-60.348518</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">4</th>
<td>-2.844254</td>
<td>-3.322700</td>
<td>-3.681745</td>
<td>-1.766435</td>
<td>-2.666579</td>
<td>-3.748774</td>
<td>-4.083619</td>
<td>-3.912834</td>
<td>-3.724181</td>
<td>-3.948160</td>
<td>...</td>
<td>-35.824612</td>
<td>-87.983566</td>
<td>-83.312317</td>
<td>-107.162407</td>
<td>-61.478374</td>
<td>-85.793571</td>
<td>-43.738819</td>
<td>-47.004211</td>
<td>-42.281624</td>
<td>-59.518513</td>
</tr>
</tbody>
</table>

<p>5 rows × 328 columns</p>
</div>
</div>
</div>
<div id="f43f61c8" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>results.iloc[<span class="dv">0</span>].sort_values(ascending<span class="op">=</span><span class="va">False</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>TLK2        8.264621
GCN2        8.101542
TLK1        7.693897
HRI         6.691402
PLK3        6.579368
             ...    
NIK       -64.605148
SRPK2     -67.300667
GAK       -69.165062
BRAF      -87.763214
PKMYT1   -113.698143
Name: 0, Length: 328, dtype: float32</code></pre>
</div>
</div>
</section>
</section>
<section id="dataset" class="level2">
<h2 class="anchored" data-anchor-id="dataset">Dataset</h2>
<p>Besides calculating sequence scores, we also provides multiple datasets of phosphorylation sites.</p>
<section id="cptac-pan-cancer-phosphoproteomics" class="level3">
<h3 class="anchored" data-anchor-id="cptac-pan-cancer-phosphoproteomics">CPTAC pan-cancer phosphoproteomics</h3>
<div id="e60c6415" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> Data.get_cptac_ensembl_site()</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>df.head(<span class="dv">3</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">gene</th>
<th data-quarto-table-cell-role="th">site</th>
<th data-quarto-table-cell-role="th">site_seq</th>
<th data-quarto-table-cell-role="th">protein</th>
<th data-quarto-table-cell-role="th">gene_name</th>
<th data-quarto-table-cell-role="th">gene_site</th>
<th data-quarto-table-cell-role="th">protein_site</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">0</th>
<td>ENSG00000003056.8</td>
<td>S267</td>
<td>DDQLGEESEERDDHL</td>
<td>ENSP00000000412.3</td>
<td>M6PR</td>
<td>M6PR_S267</td>
<td>ENSP00000000412_S267</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>ENSG00000003056.8</td>
<td>S267</td>
<td>DDQLGEESEERDDHL</td>
<td>ENSP00000440488.2</td>
<td>M6PR</td>
<td>M6PR_S267</td>
<td>ENSP00000440488_S267</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>ENSG00000048028.11</td>
<td>S1053</td>
<td>PPTIRPNSPYDLCSR</td>
<td>ENSP00000003302.4</td>
<td>USP28</td>
<td>USP28_S1053</td>
<td>ENSP00000003302_S1053</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="ochoa-et-al.-human-phosphoproteome" class="level3">
<h3 class="anchored" data-anchor-id="ochoa-et-al.-human-phosphoproteome"><a href="https://www.nature.com/articles/s41587-019-0344-3">Ochoa et al.&nbsp;human phosphoproteome</a></h3>
<div id="aa48dbe7" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> Data.get_ochoa_site()</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>df.head(<span class="dv">3</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">uniprot</th>
<th data-quarto-table-cell-role="th">position</th>
<th data-quarto-table-cell-role="th">residue</th>
<th data-quarto-table-cell-role="th">is_disopred</th>
<th data-quarto-table-cell-role="th">disopred_score</th>
<th data-quarto-table-cell-role="th">log10_hotspot_pval_min</th>
<th data-quarto-table-cell-role="th">isHotspot</th>
<th data-quarto-table-cell-role="th">uniprot_position</th>
<th data-quarto-table-cell-role="th">functional_score</th>
<th data-quarto-table-cell-role="th">current_uniprot</th>
<th data-quarto-table-cell-role="th">name</th>
<th data-quarto-table-cell-role="th">gene</th>
<th data-quarto-table-cell-role="th">Sequence</th>
<th data-quarto-table-cell-role="th">is_valid</th>
<th data-quarto-table-cell-role="th">site_seq</th>
<th data-quarto-table-cell-role="th">gene_site</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">0</th>
<td>A0A075B6Q4</td>
<td>24</td>
<td>S</td>
<td>1.0</td>
<td>0.91</td>
<td>6.839384</td>
<td>1.0</td>
<td>A0A075B6Q4_24</td>
<td>0.149257</td>
<td>A0A075B6Q4</td>
<td>A0A075B6Q4_HUMAN</td>
<td>None</td>
<td>MDIQKSENEDDSEWEDVDDEKGDSNDDYDSAGLLSDEDCMSVPGKT...</td>
<td>True</td>
<td>VDDEKGDSNDDYDSA</td>
<td>A0A075B6Q4_S24</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>A0A075B6Q4</td>
<td>35</td>
<td>S</td>
<td>1.0</td>
<td>0.87</td>
<td>9.192622</td>
<td>0.0</td>
<td>A0A075B6Q4_35</td>
<td>0.136966</td>
<td>A0A075B6Q4</td>
<td>A0A075B6Q4_HUMAN</td>
<td>None</td>
<td>MDIQKSENEDDSEWEDVDDEKGDSNDDYDSAGLLSDEDCMSVPGKT...</td>
<td>True</td>
<td>YDSAGLLSDEDCMSV</td>
<td>A0A075B6Q4_S35</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>A0A075B6Q4</td>
<td>57</td>
<td>S</td>
<td>0.0</td>
<td>0.28</td>
<td>0.818834</td>
<td>0.0</td>
<td>A0A075B6Q4_57</td>
<td>0.125364</td>
<td>A0A075B6Q4</td>
<td>A0A075B6Q4_HUMAN</td>
<td>None</td>
<td>MDIQKSENEDDSEWEDVDDEKGDSNDDYDSAGLLSDEDCMSVPGKT...</td>
<td>True</td>
<td>IADHLFWSEETKSRF</td>
<td>A0A075B6Q4_S57</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="phosphositeplus-human-phosphorylation-site" class="level3">
<h3 class="anchored" data-anchor-id="phosphositeplus-human-phosphorylation-site">PhosphoSitePlus human phosphorylation site</h3>
<div id="30a45a16" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> Data.get_psp_human_site()</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>df.head(<span class="dv">3</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">gene</th>
<th data-quarto-table-cell-role="th">protein</th>
<th data-quarto-table-cell-role="th">uniprot</th>
<th data-quarto-table-cell-role="th">site</th>
<th data-quarto-table-cell-role="th">gene_site</th>
<th data-quarto-table-cell-role="th">SITE_GRP_ID</th>
<th data-quarto-table-cell-role="th">species</th>
<th data-quarto-table-cell-role="th">site_seq</th>
<th data-quarto-table-cell-role="th">LT_LIT</th>
<th data-quarto-table-cell-role="th">MS_LIT</th>
<th data-quarto-table-cell-role="th">MS_CST</th>
<th data-quarto-table-cell-role="th">CST_CAT#</th>
<th data-quarto-table-cell-role="th">Ambiguous_Site</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">0</th>
<td>YWHAB</td>
<td>14-3-3 beta</td>
<td>P31946</td>
<td>T2</td>
<td>YWHAB_T2</td>
<td>15718712</td>
<td>human</td>
<td>______MtMDksELV</td>
<td>NaN</td>
<td>3.0</td>
<td>1.0</td>
<td>None</td>
<td>0</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>YWHAB</td>
<td>14-3-3 beta</td>
<td>P31946</td>
<td>S6</td>
<td>YWHAB_S6</td>
<td>15718709</td>
<td>human</td>
<td>__MtMDksELVQkAk</td>
<td>NaN</td>
<td>8.0</td>
<td>NaN</td>
<td>None</td>
<td>0</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>YWHAB</td>
<td>14-3-3 beta</td>
<td>P31946</td>
<td>Y21</td>
<td>YWHAB_Y21</td>
<td>3426383</td>
<td>human</td>
<td>LAEQAERyDDMAAAM</td>
<td>NaN</td>
<td>NaN</td>
<td>4.0</td>
<td>None</td>
<td>0</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="unique-sites-of-combined-ochoa-phosphositeplus" class="level3">
<h3 class="anchored" data-anchor-id="unique-sites-of-combined-ochoa-phosphositeplus">Unique sites of combined Ochoa &amp; PhosphoSitePlus</h3>
<div id="b15ca02d" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> Data.get_combine_site_psp_ochoa()</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>df.head(<span class="dv">3</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">uniprot</th>
<th data-quarto-table-cell-role="th">gene</th>
<th data-quarto-table-cell-role="th">site</th>
<th data-quarto-table-cell-role="th">site_seq</th>
<th data-quarto-table-cell-role="th">source</th>
<th data-quarto-table-cell-role="th">AM_pathogenicity</th>
<th data-quarto-table-cell-role="th">CDDM_upper</th>
<th data-quarto-table-cell-role="th">CDDM_max_score</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">0</th>
<td>A0A024R4G9</td>
<td>C19orf48</td>
<td>S20</td>
<td>ITGSRLLSMVPGPAR</td>
<td>psp</td>
<td>NaN</td>
<td>PRKX,AKT1,PKG1,P90RSK,HIPK4,AKT3,HIPK1,PKACB,H...</td>
<td>2.407041</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>A0A075B6Q4</td>
<td>None</td>
<td>S24</td>
<td>VDDEKGDSNDDYDSA</td>
<td>ochoa</td>
<td>NaN</td>
<td>CK2A2,CK2A1,GRK7,GRK5,CK1G1,CK1A,IKKA,CK1G2,CA...</td>
<td>2.295654</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>A0A075B6Q4</td>
<td>None</td>
<td>S35</td>
<td>YDSAGLLSDEDCMSV</td>
<td>ochoa</td>
<td>NaN</td>
<td>CK2A2,CK2A1,IKKA,ATM,IKKB,CAMK1D,MARK2,GRK7,IK...</td>
<td>2.488683</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
</section>
<section id="phosphorylation-site-sequence-example" class="level2">
<h2 class="anchored" data-anchor-id="phosphorylation-site-sequence-example">Phosphorylation site sequence example</h2>
<p><strong><em>All capital - 15 length (-7 to +7)</em></strong></p>
<ul>
<li>QSEEEKLSPSPTTED</li>
<li>TLQHVPDYRQNVYIP</li>
<li>TMGLSARyGPQFTLQ</li>
</ul>
<p><strong><em>All capital - 10 length (-5 to +4)</em></strong></p>
<ul>
<li>SRDPHYQDPH</li>
<li>LDNPDyQQDF</li>
<li>AAAAAsGGAG</li>
</ul>
<p><strong><em>With lowercase - (-7 to +7)</em></strong></p>
<ul>
<li>QsEEEKLsPsPTTED</li>
<li>TLQHVPDyRQNVYIP</li>
<li>TMGLsARyGPQFTLQ</li>
</ul>
<p><strong><em>With lowercase - (-5 to +4)</em></strong></p>
<ul>
<li>sRDPHyQDPH</li>
<li>LDNPDyQQDF</li>
<li>AAAAAsGGAG</li>
</ul>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/sky1ove\.github\.io\/katlas");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/sky1ove/katlas/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>