<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.27">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Pathway analysis – python-katlas</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdn.jsdelivr.net/npm/jquery@3.5.1/dist/jquery.min.js" integrity="sha384-ZvpUoO/+PpLXR1lu4jmpXWu80pZlYUAfxl5NsBMWOEPSjUn/6Z/hRTt8+pR6L4N2" crossorigin="anonymous"></script><script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-ed96de9b727972fe78a7b5d16c58bf87.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-45a6b1ee3682cc1dbb0c772b170fe9b7.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdn.jsdelivr.net/npm/requirejs@2.3.6/require.min.js" integrity="sha384-c9c+LnTbwQ3aujuU7ULEPVvgLs+Fn6fJUvIGTsuu1ZcCf11fiEubah0ttpca4ntM sha384-6V1/AdqZRWk1KAlWbKBlGhN7VG4iE/yAZcO6NZPMF8od0vukrvr0tg4qY6NSrItx" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="Pathway analysis – python-katlas">
<meta property="og:description" content="tools for predicting kinome specificities">
<meta property="og:image" content="https://sky1ove.github.io/katlas/07_pathway_files/figure-html/cell-18-output-1.png">
<meta property="og:site_name" content="python-katlas">
<meta property="og:image:height" content="312">
<meta property="og:image:width" content="682">
<meta name="twitter:title" content="Pathway analysis – python-katlas">
<meta name="twitter:description" content="tools for predicting kinome specificities">
<meta name="twitter:image" content="https://sky1ove.github.io/katlas/07_pathway_files/figure-html/cell-18-output-1.png">
<meta name="twitter:image-height" content="312">
<meta name="twitter:image-width" content="682">
<meta name="twitter:card" content="summary_large_image">
</head>

<body class="nav-sidebar floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">python-katlas</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/sky1ove/katlas"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./data.html">Modules</a></li><li class="breadcrumb-item"><a href="./pathway.html">Pathway analysis</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Get Started</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Modules</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./utils.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Utils</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pssm_core.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">pssm.core</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pssm_plot.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">pssm.plot</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pssm_lo.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">pssm.lo</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pssm_pspa.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">pssm.pspa</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pssm_compare.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">pssm.compare</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./scoring.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Scoring</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Hierarchical clustering</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./hierarchical.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Hierarchical clustering</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./feature.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Feature</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./plot.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Plot</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pathway.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Pathway analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ml.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Train ML</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./dnn.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Train DL</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#overview" id="toc-overview" class="nav-link active" data-scroll-target="#overview">Overview</a></li>
  <li><a href="#setup" id="toc-setup" class="nav-link" data-scroll-target="#setup">Setup</a></li>
  <li><a href="#reactome-pathway" id="toc-reactome-pathway" class="nav-link" data-scroll-target="#reactome-pathway">Reactome pathway</a>
  <ul class="collapse">
  <li><a href="#get_reactome_raw" id="toc-get_reactome_raw" class="nav-link" data-scroll-target="#get_reactome_raw">get_reactome_raw</a></li>
  <li><a href="#get_reactome" id="toc-get_reactome" class="nav-link" data-scroll-target="#get_reactome">get_reactome</a></li>
  </ul></li>
  <li><a href="#reference" id="toc-reference" class="nav-link" data-scroll-target="#reference">Reference</a>
  <ul class="collapse">
  <li><a href="#query_reactome" id="toc-query_reactome" class="nav-link" data-scroll-target="#query_reactome">query_reactome</a></li>
  <li><a href="#add_reactome_ref" id="toc-add_reactome_ref" class="nav-link" data-scroll-target="#add_reactome_ref">add_reactome_ref</a></li>
  </ul></li>
  <li><a href="#bar-plot-of-pathways" id="toc-bar-plot-of-pathways" class="nav-link" data-scroll-target="#bar-plot-of-pathways">Bar plot of pathways</a>
  <ul class="collapse">
  <li><a href="#plot_path" id="toc-plot_path" class="nav-link" data-scroll-target="#plot_path">plot_path</a></li>
  </ul></li>
  <li><a href="#overlap" id="toc-overlap" class="nav-link" data-scroll-target="#overlap">Overlap</a>
  <ul class="collapse">
  <li><a href="#get_overlap" id="toc-get_overlap" class="nav-link" data-scroll-target="#get_overlap">get_overlap</a></li>
  </ul></li>
  <li><a href="#pipeline" id="toc-pipeline" class="nav-link" data-scroll-target="#pipeline">Pipeline</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/sky1ove/katlas/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./data.html">Modules</a></li><li class="breadcrumb-item"><a href="./pathway.html">Pathway analysis</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">Pathway analysis</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<section id="overview" class="level2">
<h2 class="anchored" data-anchor-id="overview">Overview</h2>
<p><strong>Reactome Pathway Enrichment</strong></p>
<p><code>get_reactome_raw(gene_list)</code> — Performs Reactome pathway analysis on a gene set and returns the complete raw output as a DataFrame. Includes all pathway metadata (IDs, names, p-values, entity counts, reaction ratios).</p>
<p><code>get_reactome(gene_list, p_type)</code> — Wrapper around <code>get_reactome_raw</code> that returns a cleaned DataFrame with pathway names, Reactome IDs, p-values (or FDR), and -log10 transformed values.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>path <span class="op">=</span> get_reactome(</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>    gene_list<span class="op">=</span>pi3ks,   <span class="co"># list of gene symbols to analyze</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>    p_type<span class="op">=</span><span class="st">'FDR'</span>,      <span class="co"># 'FDR' or 'p' for raw p-value</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<hr>
<p><strong>Reference Database Queries</strong></p>
<p><code>query_reactome(uniprot_id)</code> — Queries a UniProt ID against the Reactome pathway database. Returns all pathways (at any hierarchy level) that the protein participates in, with a <code>lowest</code> flag indicating leaf-level pathways.</p>
<p><code>add_reactome_ref(df, uniprot)</code> — Augments a pathway analysis DataFrame with reference columns indicating whether each pathway is annotated for a specific protein. Also appends any missing pathways from the reference (with p=1.0).</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>out <span class="op">=</span> add_reactome_ref(</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>    df<span class="op">=</span>path,           <span class="co"># output DataFrame from get_reactome()</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    uniprot<span class="op">=</span><span class="st">'P31749'</span>,  <span class="co"># UniProt ID of kinase to check against</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Adds columns: ref_path (1/0), ref_path_lowest (1/0)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<hr>
<p><strong>Visualization</strong></p>
<p><code>plot_path(react_df, ...)</code> — Creates a horizontal bar plot of top enriched pathways. Optionally highlights bars in dark red for pathways matching a reference set.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>plot_path(</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>    react_df<span class="op">=</span>out,                <span class="co"># pathway DataFrame with -log10 p-values</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    p_type<span class="op">=</span><span class="st">'FDR'</span>,                <span class="co"># 'FDR' or 'p' - which p-value column to plot</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    ref_col<span class="op">=</span><span class="st">'ref_path'</span>,          <span class="co"># column name (1/0) to highlight matching bars</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    ref_id_list<span class="op">=</span><span class="va">None</span>,            <span class="co"># alternative: list of reactome_ids to highlight</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>    top_n<span class="op">=</span><span class="dv">15</span>,                    <span class="co"># number of top pathways to display</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>    max_label_length<span class="op">=</span><span class="dv">80</span>,         <span class="co"># truncate pathway names longer than this</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<hr>
<p><strong>Overlap Statistics</strong></p>
<p><code>get_overlap(react_df, ...)</code> — Calculates what fraction of reference pathways pass a significance threshold. Optionally plots a histogram of -log10(p) values with a threshold line.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>accuracy <span class="op">=</span> get_overlap(</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>    react_df<span class="op">=</span>out,                <span class="co"># pathway DataFrame from add_reactome_ref()</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    ref_col<span class="op">=</span><span class="st">'ref_path_lowest'</span>,   <span class="co"># column name (1/0) indicating reference pathways</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    ref_id_list<span class="op">=</span><span class="va">None</span>,            <span class="co"># alternative: list of reactome_ids</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    p_type<span class="op">=</span><span class="st">'FDR'</span>,                <span class="co"># 'FDR' or 'p'</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    thr<span class="op">=</span><span class="fl">0.05</span>,                    <span class="co"># significance threshold (before log transform)</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>    plot<span class="op">=</span><span class="va">True</span>,                   <span class="co"># whether to show histogram</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>    figsize<span class="op">=</span>(<span class="dv">5</span>, <span class="dv">3</span>),              <span class="co"># figure size tuple</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Returns: float (fraction passing threshold, e.g., 0.97)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<hr>
<p><strong>Complete Pipeline Example</strong></p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># 1. Run pathway enrichment on gene set</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>out <span class="op">=</span> get_reactome(</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    gene_list<span class="op">=</span>pi3ks,   <span class="co"># list of gene symbols</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    p_type<span class="op">=</span><span class="st">'FDR'</span>,      <span class="co"># use FDR-corrected p-values</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co"># 2. Add reference annotations for a specific kinase</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>out <span class="op">=</span> add_reactome_ref(</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>    df<span class="op">=</span>out,            <span class="co"># enrichment results</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>    uniprot<span class="op">=</span><span class="st">'P31749'</span>,  <span class="co"># AKT1 UniProt ID</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a><span class="co"># 3. Visualize top pathways with reference highlighting</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>plot_path(</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>    react_df<span class="op">=</span>out,              <span class="co"># augmented DataFrame</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>    p_type<span class="op">=</span><span class="st">'FDR'</span>,              <span class="co"># plot -log10(FDR)</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>    ref_col<span class="op">=</span><span class="st">'ref_path_lowest'</span>, <span class="co"># highlight lowest-level reference pathways</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>    top_n<span class="op">=</span><span class="dv">15</span>,                  <span class="co"># show top 15</span></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a><span class="co"># 4. Calculate overlap accuracy</span></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>accuracy <span class="op">=</span> get_overlap(</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>    react_df<span class="op">=</span>out,              <span class="co"># augmented DataFrame</span></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>    ref_col<span class="op">=</span><span class="st">'ref_path_lowest'</span>, <span class="co"># evaluate against lowest-level pathways</span></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>    thr<span class="op">=</span><span class="fl">0.05</span>,                  <span class="co"># FDR &lt; 0.05 significance cutoff</span></span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>    plot<span class="op">=</span><span class="va">True</span>,                 <span class="co"># show histogram</span></span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="setup" class="level2">
<h2 class="anchored" data-anchor-id="setup">Setup</h2>
</section>
<section id="reactome-pathway" class="level2">
<h2 class="anchored" data-anchor-id="reactome-pathway">Reactome pathway</h2>
<hr>
<section id="get_reactome_raw" class="level3">
<h3 class="anchored" data-anchor-id="get_reactome_raw">get_reactome_raw</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_reactome_raw(</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>    gene_list</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>):</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Reactome pathway analysis for a given gene set; returns raw output in dataframe.</em></p>
<div id="4be3bf52-1d0f-46d3-ad05-1b0689d5ff2c" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>pi3ks<span class="op">=</span>[<span class="st">'PIK3CA'</span>,<span class="st">'PIK3CB'</span>,<span class="st">'PIK3CD'</span>,<span class="st">'PIK3CG'</span>,<span class="st">'PIK3R1'</span>,<span class="st">'PIK3R2'</span>,<span class="st">'PIK3R3'</span>,<span class="st">'PTEN'</span>,<span class="st">'AKT1'</span>,<span class="st">'AKT2'</span>,<span class="st">'AKT3'</span>,<span class="st">'MTOR'</span>,<span class="st">'RICTOR'</span>,<span class="st">'RPTOR'</span>,<span class="st">'TSC1'</span>,<span class="st">'TSC2'</span>,<span class="st">'PDK1'</span>,<span class="st">'IRS1'</span>,<span class="st">'IRS2'</span>,<span class="st">'INSR'</span>,<span class="st">'IGF1R'</span>,<span class="st">'GAB1'</span>,<span class="st">'HRAS'</span>,<span class="st">'NRAS'</span>,<span class="st">'KRAS'</span>,<span class="st">'EGFR'</span>,<span class="st">'ERBB2'</span>,<span class="st">'ERBB3'</span>,<span class="st">'ERBB4'</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="6ca3d3c6-5de4-464e-b315-e59989f490ba" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>raw_out <span class="op">=</span> get_reactome_raw(pi3ks)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>raw_out.head()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">stId</th>
<th data-quarto-table-cell-role="th">dbId</th>
<th data-quarto-table-cell-role="th">name</th>
<th data-quarto-table-cell-role="th">llp</th>
<th data-quarto-table-cell-role="th">inDisease</th>
<th data-quarto-table-cell-role="th">species.dbId</th>
<th data-quarto-table-cell-role="th">species.taxId</th>
<th data-quarto-table-cell-role="th">species.name</th>
<th data-quarto-table-cell-role="th">entities.resource</th>
<th data-quarto-table-cell-role="th">entities.total</th>
<th data-quarto-table-cell-role="th">entities.found</th>
<th data-quarto-table-cell-role="th">entities.ratio</th>
<th data-quarto-table-cell-role="th">entities.pValue</th>
<th data-quarto-table-cell-role="th">entities.fdr</th>
<th data-quarto-table-cell-role="th">entities.exp</th>
<th data-quarto-table-cell-role="th">reactions.resource</th>
<th data-quarto-table-cell-role="th">reactions.total</th>
<th data-quarto-table-cell-role="th">reactions.found</th>
<th data-quarto-table-cell-role="th">reactions.ratio</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">0</th>
<td>R-HSA-1963640</td>
<td>1963640</td>
<td>GRB2 events in ERBB2 signaling</td>
<td>True</td>
<td>False</td>
<td>48887</td>
<td>9606</td>
<td>Homo sapiens</td>
<td>TOTAL</td>
<td>21</td>
<td>9</td>
<td>0.001301</td>
<td>1.110223e-16</td>
<td>1.221245e-15</td>
<td>[]</td>
<td>TOTAL</td>
<td>4</td>
<td>4</td>
<td>0.000252</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>R-HSA-9665348</td>
<td>9665348</td>
<td>Signaling by ERBB2 ECD mutants</td>
<td>True</td>
<td>True</td>
<td>48887</td>
<td>9606</td>
<td>Homo sapiens</td>
<td>TOTAL</td>
<td>23</td>
<td>9</td>
<td>0.001425</td>
<td>1.110223e-16</td>
<td>1.221245e-15</td>
<td>[]</td>
<td>TOTAL</td>
<td>15</td>
<td>15</td>
<td>0.000945</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>R-HSA-9664565</td>
<td>9664565</td>
<td>Signaling by ERBB2 KD Mutants</td>
<td>True</td>
<td>True</td>
<td>48887</td>
<td>9606</td>
<td>Homo sapiens</td>
<td>TOTAL</td>
<td>35</td>
<td>13</td>
<td>0.002168</td>
<td>1.110223e-16</td>
<td>1.221245e-15</td>
<td>[]</td>
<td>TOTAL</td>
<td>17</td>
<td>17</td>
<td>0.001071</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">3</th>
<td>R-HSA-1227990</td>
<td>1227990</td>
<td>Signaling by ERBB2 in Cancer</td>
<td>False</td>
<td>True</td>
<td>48887</td>
<td>9606</td>
<td>Homo sapiens</td>
<td>TOTAL</td>
<td>36</td>
<td>13</td>
<td>0.002230</td>
<td>1.110223e-16</td>
<td>1.221245e-15</td>
<td>[]</td>
<td>TOTAL</td>
<td>62</td>
<td>62</td>
<td>0.003907</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">4</th>
<td>R-HSA-9665686</td>
<td>9665686</td>
<td>Signaling by ERBB2 TMD/JMD mutants</td>
<td>True</td>
<td>True</td>
<td>48887</td>
<td>9606</td>
<td>Homo sapiens</td>
<td>TOTAL</td>
<td>30</td>
<td>10</td>
<td>0.001858</td>
<td>1.110223e-16</td>
<td>1.221245e-15</td>
<td>[]</td>
<td>TOTAL</td>
<td>13</td>
<td>13</td>
<td>0.000819</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<hr>
</section>
<section id="get_reactome" class="level3">
<h3 class="anchored" data-anchor-id="get_reactome">get_reactome</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_reactome(</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>    gene_list, p_type:<span class="bu">str</span><span class="op">=</span><span class="st">'FDR'</span>, <span class="co"># or p</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>):</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Reactome pathway analysis for a given gene set; returns formated output in dataframe with additional -log10(p)</em></p>
<div id="a42e6a3a-4ca2-4ea2-b7c5-ccdd5d978412" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>path <span class="op">=</span> get_reactome(pi3ks,p_type<span class="op">=</span><span class="st">'p'</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>path.head()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Running pathway analysis
Done</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">name</th>
<th data-quarto-table-cell-role="th">reactome_id</th>
<th data-quarto-table-cell-role="th">p</th>
<th data-quarto-table-cell-role="th">-log10_p</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">0</th>
<td>GRB2 events in ERBB2 signaling</td>
<td>R-HSA-1963640</td>
<td>1.110223e-16</td>
<td>15.955</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>Signaling by ERBB2 ECD mutants</td>
<td>R-HSA-9665348</td>
<td>1.110223e-16</td>
<td>15.955</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>Signaling by ERBB2 KD Mutants</td>
<td>R-HSA-9664565</td>
<td>1.110223e-16</td>
<td>15.955</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">3</th>
<td>Signaling by ERBB2 in Cancer</td>
<td>R-HSA-1227990</td>
<td>1.110223e-16</td>
<td>15.955</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">4</th>
<td>Signaling by ERBB2 TMD/JMD mutants</td>
<td>R-HSA-9665686</td>
<td>1.110223e-16</td>
<td>15.955</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="d5e6d4d9-f5a1-463c-812f-b4aa4c3bd5c5" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>path <span class="op">=</span> get_reactome(pi3ks,p_type<span class="op">=</span><span class="st">'FDR'</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>path.head()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Running pathway analysis
Done</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">name</th>
<th data-quarto-table-cell-role="th">reactome_id</th>
<th data-quarto-table-cell-role="th">FDR</th>
<th data-quarto-table-cell-role="th">-log10_FDR</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">0</th>
<td>GRB2 events in ERBB2 signaling</td>
<td>R-HSA-1963640</td>
<td>1.221245e-15</td>
<td>14.913</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>Signaling by ERBB2 ECD mutants</td>
<td>R-HSA-9665348</td>
<td>1.221245e-15</td>
<td>14.913</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>Signaling by ERBB2 KD Mutants</td>
<td>R-HSA-9664565</td>
<td>1.221245e-15</td>
<td>14.913</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">3</th>
<td>Signaling by ERBB2 in Cancer</td>
<td>R-HSA-1227990</td>
<td>1.221245e-15</td>
<td>14.913</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">4</th>
<td>Signaling by ERBB2 TMD/JMD mutants</td>
<td>R-HSA-9665686</td>
<td>1.221245e-15</td>
<td>14.913</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="dee1654f-87fa-402d-af5a-bd63c3574cbe" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>path[path.FDR<span class="op">&lt;</span><span class="fl">0.05</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">name</th>
<th data-quarto-table-cell-role="th">reactome_id</th>
<th data-quarto-table-cell-role="th">FDR</th>
<th data-quarto-table-cell-role="th">-log10_FDR</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">0</th>
<td>GRB2 events in ERBB2 signaling</td>
<td>R-HSA-1963640</td>
<td>1.221245e-15</td>
<td>14.913</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>Signaling by ERBB2 ECD mutants</td>
<td>R-HSA-9665348</td>
<td>1.221245e-15</td>
<td>14.913</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>Signaling by ERBB2 KD Mutants</td>
<td>R-HSA-9664565</td>
<td>1.221245e-15</td>
<td>14.913</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">3</th>
<td>Signaling by ERBB2 in Cancer</td>
<td>R-HSA-1227990</td>
<td>1.221245e-15</td>
<td>14.913</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">4</th>
<td>Signaling by ERBB2 TMD/JMD mutants</td>
<td>R-HSA-9665686</td>
<td>1.221245e-15</td>
<td>14.913</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">...</th>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">324</th>
<td>Signaling by Rho GTPases</td>
<td>R-HSA-194315</td>
<td>4.264432e-02</td>
<td>1.370</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">325</th>
<td>RUNX3 regulates p14-ARF</td>
<td>R-HSA-8951936</td>
<td>4.269110e-02</td>
<td>1.370</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">326</th>
<td>Cellular response to chemical stress</td>
<td>R-HSA-9711123</td>
<td>4.373732e-02</td>
<td>1.359</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">327</th>
<td>Signaling by Rho GTPases, Miro GTPases and RHO...</td>
<td>R-HSA-9716542</td>
<td>4.619382e-02</td>
<td>1.335</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">328</th>
<td>Metabolism of lipids</td>
<td>R-HSA-556833</td>
<td>4.924793e-02</td>
<td>1.308</td>
</tr>
</tbody>
</table>

<p>329 rows × 4 columns</p>
</div>
</div>
</div>
</section>
</section>
<section id="reference" class="level2">
<h2 class="anchored" data-anchor-id="reference">Reference</h2>
<p>Download from Reactome/Download_data: https://reactome.org/download-data</p>
<p>Download <code>UniProt to All pathways</code> under <code>Identifier mapping files</code></p>
<p>for type, there are IEA (Inferred from Electronic Annotation) and TAS (Traceable Author Statement, higher confidence)</p>
<div id="8f72c970-d2b5-4388-89d9-91eb3315e7a4" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>ref <span class="op">=</span> Data.get_reactome_pathway()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="d1b095b6-61e4-44a5-b36c-e23e62f87ac3" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>ref.head()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">uniprot</th>
<th data-quarto-table-cell-role="th">reactome_id</th>
<th data-quarto-table-cell-role="th">pathway</th>
<th data-quarto-table-cell-role="th">type</th>
<th data-quarto-table-cell-role="th">species</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">0</th>
<td>A0A023GPK8</td>
<td>R-DME-1500931</td>
<td>Cell-Cell communication</td>
<td>IEA</td>
<td>Drosophila melanogaster</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>A0A023GPK8</td>
<td>R-DME-373753</td>
<td>Nephrin family interactions</td>
<td>IEA</td>
<td>Drosophila melanogaster</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>A0A023GRW3</td>
<td>R-DME-72163</td>
<td>mRNA Splicing - Major Pathway</td>
<td>IEA</td>
<td>Drosophila melanogaster</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">3</th>
<td>A0A023GRW3</td>
<td>R-DME-72172</td>
<td>mRNA Splicing</td>
<td>IEA</td>
<td>Drosophila melanogaster</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">4</th>
<td>A0A023GRW3</td>
<td>R-DME-72203</td>
<td>Processing of Capped Intron-Containing Pre-mRNA</td>
<td>IEA</td>
<td>Drosophila melanogaster</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<hr>
<section id="query_reactome" class="level3">
<h3 class="anchored" data-anchor-id="query_reactome">query_reactome</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> query_reactome(</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>    uniprot_id</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>):</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Query uniprot ID in Reactome all level pathway database.</em></p>
<div id="ea287611-ec66-47ed-82fb-cc0b8d947bd0" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>uniprot<span class="op">=</span><span class="st">'P31751'</span> <span class="co"># AKT2</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>akt_path <span class="op">=</span> query_reactome(uniprot) <span class="co"># AKT2</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="117a44b7-000d-474b-a558-38328cc9da0f" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>akt_path</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">reactome_id</th>
<th data-quarto-table-cell-role="th">uniprot</th>
<th data-quarto-table-cell-role="th">pathway</th>
<th data-quarto-table-cell-role="th">type</th>
<th data-quarto-table-cell-role="th">species</th>
<th data-quarto-table-cell-role="th">lowest</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">0</th>
<td>R-HSA-109581</td>
<td>P31751</td>
<td>Apoptosis</td>
<td>IEA</td>
<td>Homo sapiens</td>
<td>0</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>R-HSA-109606</td>
<td>P31751</td>
<td>Intrinsic Pathway for Apoptosis</td>
<td>IEA</td>
<td>Homo sapiens</td>
<td>0</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>R-HSA-109703</td>
<td>P31751</td>
<td>PKB-mediated events</td>
<td>IEA</td>
<td>Homo sapiens</td>
<td>0</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">3</th>
<td>R-HSA-109704</td>
<td>P31751</td>
<td>PI3K Cascade</td>
<td>IEA, TAS</td>
<td>Homo sapiens</td>
<td>0</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">4</th>
<td>R-HSA-111447</td>
<td>P31751</td>
<td>Activation of BAD and translocation to mitocho...</td>
<td>IEA</td>
<td>Homo sapiens</td>
<td>1</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">...</th>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">93</th>
<td>R-HSA-9755511</td>
<td>P31751</td>
<td>KEAP1-NFE2L2 pathway</td>
<td>TAS</td>
<td>Homo sapiens</td>
<td>1</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">94</th>
<td>R-HSA-9755779</td>
<td>P31751</td>
<td>SARS-CoV-2 targets host intracellular signalli...</td>
<td>TAS</td>
<td>Homo sapiens</td>
<td>1</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">95</th>
<td>R-HSA-9824446</td>
<td>P31751</td>
<td>Viral Infection Pathways</td>
<td>TAS</td>
<td>Homo sapiens</td>
<td>0</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">96</th>
<td>R-HSA-9824585</td>
<td>P31751</td>
<td>Regulation of MITF-M-dependent genes involved ...</td>
<td>IEA</td>
<td>Homo sapiens</td>
<td>1</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">97</th>
<td>R-HSA-9856651</td>
<td>P31751</td>
<td>MITF-M-dependent gene expression</td>
<td>IEA</td>
<td>Homo sapiens</td>
<td>0</td>
</tr>
</tbody>
</table>

<p>98 rows × 6 columns</p>
</div>
</div>
</div>
<div id="8caa6784-bc39-43e3-8bfa-d1e80024492f" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># lowest</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>akt_path[akt_path.lowest<span class="op">==</span><span class="dv">1</span>].shape</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>(32, 6)</code></pre>
</div>
</div>
<hr>
</section>
<section id="add_reactome_ref" class="level3">
<h3 class="anchored" data-anchor-id="add_reactome_ref">add_reactome_ref</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> add_reactome_ref(</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>    df, uniprot</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>):</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div id="92351a6f" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>out <span class="op">=</span> add_reactome_ref(path,uniprot)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>out.shape</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>(406, 6)</code></pre>
</div>
</div>
</section>
</section>
<section id="bar-plot-of-pathways" class="level2">
<h2 class="anchored" data-anchor-id="bar-plot-of-pathways">Bar plot of pathways</h2>
<hr>
<section id="plot_path" class="level3">
<h3 class="anchored" data-anchor-id="plot_path">plot_path</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_path(</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>    react_df, <span class="co"># the output df of get_reactome</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>    p_type:<span class="bu">str</span><span class="op">=</span><span class="st">'FDR'</span>, ref_id_list:NoneType<span class="op">=</span><span class="va">None</span>, <span class="co"># list of reactome_id</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>    ref_col:NoneType<span class="op">=</span><span class="va">None</span>, <span class="co"># column in reac_df, 1 or 0 to indicate whether it's in ref</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>    top_n:<span class="bu">int</span><span class="op">=</span><span class="dv">10</span>, max_label_length:<span class="bu">int</span><span class="op">=</span><span class="dv">80</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>):</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Plot the output of get_reactome.</em> If ref_df is provided, bars corresponding to pathways in ref_df are shown in dark red.</p>
<div id="99859ee4-c961-48be-862e-853cec5844c8" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>plot_path(out)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">'PI3K Pathways'</span>)<span class="op">;</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="07_pathway_files/figure-html/cell-18-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="9acc19ee-de50-4bf2-9e7d-2d7de0335cb4" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># All level</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>plot_path(out,p_type<span class="op">=</span><span class="st">'FDR'</span>,ref_id_list<span class="op">=</span>akt_path.reactome_id,top_n<span class="op">=</span><span class="dv">15</span>)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">'PI3K Pathways (with highlight as overlap with all level Reactome database)'</span>)<span class="op">;</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="07_pathway_files/figure-html/cell-19-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="ecd7e69d-90aa-4fb6-8359-f4c347f067c2" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>out.head()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">name</th>
<th data-quarto-table-cell-role="th">reactome_id</th>
<th data-quarto-table-cell-role="th">FDR</th>
<th data-quarto-table-cell-role="th">-log10_FDR</th>
<th data-quarto-table-cell-role="th">ref_path</th>
<th data-quarto-table-cell-role="th">ref_path_lowest</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">0</th>
<td>GRB2 events in ERBB2 signaling</td>
<td>R-HSA-1963640</td>
<td>1.221245e-15</td>
<td>14.913</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>Signaling by ERBB2 ECD mutants</td>
<td>R-HSA-9665348</td>
<td>1.221245e-15</td>
<td>14.913</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>Signaling by ERBB2 KD Mutants</td>
<td>R-HSA-9664565</td>
<td>1.221245e-15</td>
<td>14.913</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">3</th>
<td>Signaling by ERBB2 in Cancer</td>
<td>R-HSA-1227990</td>
<td>1.221245e-15</td>
<td>14.913</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">4</th>
<td>Signaling by ERBB2 TMD/JMD mutants</td>
<td>R-HSA-9665686</td>
<td>1.221245e-15</td>
<td>14.913</td>
<td>0</td>
<td>0</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="0ac82d06-f319-4a06-947e-e886194938a5" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># All level, use ref_col</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>plot_path(out,p_type<span class="op">=</span><span class="st">'FDR'</span>,ref_col<span class="op">=</span><span class="st">'ref_path'</span>,top_n<span class="op">=</span><span class="dv">15</span>)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">'PI3K Pathways (with highlight as overlap with all level Reactome database)'</span>)<span class="op">;</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="07_pathway_files/figure-html/cell-21-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="ffd1d979-bb20-4b57-a68d-c0fcc77f3b7f" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># All level</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>plot_path(out,p_type<span class="op">=</span><span class="st">'FDR'</span>,ref_col<span class="op">=</span><span class="st">'ref_path_lowest'</span>,top_n<span class="op">=</span><span class="dv">15</span>)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">'PI3K Pathways (with highlight as overlap with lowest level Reactome database)'</span>)<span class="op">;</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="07_pathway_files/figure-html/cell-22-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="overlap" class="level2">
<h2 class="anchored" data-anchor-id="overlap">Overlap</h2>
<hr>
<section id="get_overlap" class="level3">
<h3 class="anchored" data-anchor-id="get_overlap">get_overlap</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_overlap(</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>    react_df, ref_id_list:NoneType<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>    ref_col:NoneType<span class="op">=</span><span class="va">None</span>, <span class="co"># column in react_df, 1 or 0 to indicate whether it's in ref</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>    p_type:<span class="bu">str</span><span class="op">=</span><span class="st">'FDR'</span>, thr:<span class="bu">float</span><span class="op">=</span><span class="fl">0.05</span>, <span class="co"># original threshold of p value, will be log10 transformed</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>    plot:<span class="bu">bool</span><span class="op">=</span><span class="va">True</span>, figsize:<span class="bu">tuple</span><span class="op">=</span>(<span class="dv">5</span>, <span class="dv">3</span>), kwargs:VAR_KEYWORD</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>):</span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div id="d69ac318-1ee8-433c-a8ca-af2644a91c7e" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>get_overlap(out, ref_id_list<span class="op">=</span>akt_path.reactome_id,plot<span class="op">=</span><span class="va">True</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>0.9081632653061225</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="07_pathway_files/figure-html/cell-24-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="056e7ad2-4fef-4575-b53a-82c91839f438" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>get_overlap(out, ref_col<span class="op">=</span><span class="st">'ref_path'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>0.9081632653061225</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="07_pathway_files/figure-html/cell-25-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="6f2bb3e9-d65e-4ec8-8dd3-9849c419732c" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>get_overlap(out, ref_col<span class="op">=</span><span class="st">'ref_path_lowest'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>0.96875</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="07_pathway_files/figure-html/cell-26-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="pipeline" class="level2">
<h2 class="anchored" data-anchor-id="pipeline">Pipeline</h2>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>out <span class="op">=</span> get_reactome(pi3ks,p_type<span class="op">=</span><span class="st">'FDR'</span>)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>out <span class="op">=</span> add_reactome_ref(out,<span class="st">'P31749'</span>) <span class="co"># kinase uniprot</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>accuracy <span class="op">=</span> get_overlap(out, ref_col<span class="op">=</span><span class="st">'ref_path'</span>,plot<span class="op">=</span><span class="va">True</span>) <span class="co"># if lowest, change all to lo</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/sky1ove\.github\.io\/katlas");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/sky1ove/katlas/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>