<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.27">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="PSPA data visualization">

<title>pssm.pspa – python-katlas</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdn.jsdelivr.net/npm/jquery@3.5.1/dist/jquery.min.js" integrity="sha384-ZvpUoO/+PpLXR1lu4jmpXWu80pZlYUAfxl5NsBMWOEPSjUn/6Z/hRTt8+pR6L4N2" crossorigin="anonymous"></script><script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-ed96de9b727972fe78a7b5d16c58bf87.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-45a6b1ee3682cc1dbb0c772b170fe9b7.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdn.jsdelivr.net/npm/requirejs@2.3.6/require.min.js" integrity="sha384-c9c+LnTbwQ3aujuU7ULEPVvgLs+Fn6fJUvIGTsuu1ZcCf11fiEubah0ttpca4ntM sha384-6V1/AdqZRWk1KAlWbKBlGhN7VG4iE/yAZcO6NZPMF8od0vukrvr0tg4qY6NSrItx" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="pssm.pspa – python-katlas">
<meta property="og:description" content="PSPA data visualization">
<meta property="og:image" content="https://sky1ove.github.io/katlas/02d_pssm_pspa_files/figure-html/cell-5-output-1.png">
<meta property="og:site_name" content="python-katlas">
<meta property="og:image:height" content="220">
<meta property="og:image:width" content="458">
<meta name="twitter:title" content="pssm.pspa – python-katlas">
<meta name="twitter:description" content="PSPA data visualization">
<meta name="twitter:image" content="https://sky1ove.github.io/katlas/02d_pssm_pspa_files/figure-html/cell-5-output-1.png">
<meta name="twitter:image-height" content="220">
<meta name="twitter:image-width" content="458">
<meta name="twitter:card" content="summary_large_image">
</head>

<body class="nav-sidebar floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">python-katlas</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/sky1ove/katlas"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./data.html">Modules</a></li><li class="breadcrumb-item"><a href="./pssm_pspa.html">pssm.pspa</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Get Started</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Modules</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./utils.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Utils</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pssm_core.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">pssm.core</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pssm_plot.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">pssm.plot</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pssm_lo.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">pssm.lo</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pssm_pspa.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">pssm.pspa</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pssm_compare.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">pssm.compare</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./scoring.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Scoring</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Hierarchical clustering</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./hierarchical.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Hierarchical clustering</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./feature.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Feature</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./plot.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Plot</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pathway.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Pathway analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ml.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Train ML</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./dnn.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Train DL</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#overview" id="toc-overview" class="nav-link active" data-scroll-target="#overview">Overview</a></li>
  <li><a href="#setup" id="toc-setup" class="nav-link" data-scroll-target="#setup">Setup</a></li>
  <li><a href="#plot" id="toc-plot" class="nav-link" data-scroll-target="#plot">Plot</a>
  <ul class="collapse">
  <li><a href="#preprocess_pspa" id="toc-preprocess_pspa" class="nav-link" data-scroll-target="#preprocess_pspa">preprocess_pspa</a></li>
  <li><a href="#plot_logo_pspa" id="toc-plot_logo_pspa" class="nav-link" data-scroll-target="#plot_logo_pspa">plot_logo_pspa</a></li>
  <li><a href="#plot_logo_heatmap_pspa" id="toc-plot_logo_heatmap_pspa" class="nav-link" data-scroll-target="#plot_logo_heatmap_pspa">plot_logo_heatmap_pspa</a></li>
  </ul></li>
  <li><a href="#others-old-version" id="toc-others-old-version" class="nav-link" data-scroll-target="#others-old-version">Others (old version)</a>
  <ul class="collapse">
  <li><a href="#raw2norm" id="toc-raw2norm" class="nav-link" data-scroll-target="#raw2norm">raw2norm</a></li>
  <li><a href="#get_one_kinase" id="toc-get_one_kinase" class="nav-link" data-scroll-target="#get_one_kinase">get_one_kinase</a></li>
  <li><a href="#get_logo" id="toc-get_logo" class="nav-link" data-scroll-target="#get_logo">get_logo</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/sky1ove/katlas/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./data.html">Modules</a></li><li class="breadcrumb-item"><a href="./pssm_pspa.html">pssm.pspa</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">pssm.pspa</h1>
</div>

<div>
  <div class="description">
    PSPA data visualization
  </div>
</div>


<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<section id="overview" class="level2">
<h2 class="anchored" data-anchor-id="overview">Overview</h2>
<p>This module provides tools for visualizing Position-Specific Scoring Matrices (PSSMs) from Positional Scanning Peptide Array (PSPA) kinase substrate specificity data.</p>
<hr>
<p><strong>Data Preprocessing</strong></p>
<p><code>preprocess_pspa(pssm)</code> — Prepares raw PSSM data for logo visualization. It removes the duplicate ‘s’ row (since it duplicates ‘t’), renames phospho-residues to ‘pS/pT’ and ‘pY’, and applies log₂ transformation relative to the median (excluding position 0, which is handled separately via scaling).</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>pssm <span class="op">=</span> preprocess_pspa(</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>    pssm<span class="op">=</span>raw_pssm,  <span class="co"># DataFrame with amino acids as index, positions as columns</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<hr>
<p><strong>Visualization Functions</strong></p>
<p><code>plot_logo_pspa(row, title, figsize)</code> — Generates a sequence logo from a single row of PSPA data. The letter heights represent log₂(value/median), showing amino acid preferences at each position.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>plot_logo_pspa(</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>    row<span class="op">=</span>pspa.loc[<span class="st">'GSK3B'</span>],  <span class="co"># single kinase row from Data.get_pspa()</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    title<span class="op">=</span><span class="st">'GSK3B'</span>,          <span class="co"># plot title</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>    figsize<span class="op">=</span>(<span class="dv">5</span>, <span class="dv">2</span>),         <span class="co"># figure dimensions (width, height)</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><code>plot_logo_heatmap_pspa(row, title, figsize, include_zero)</code> — Creates a combined visualization with the sequence logo on top and a heatmap below, allowing detailed inspection of specificity values at each position.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>plot_logo_heatmap_pspa(</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>    row<span class="op">=</span>pspa.loc[<span class="st">'GSK3B'</span>],  <span class="co"># single kinase row from Data.get_pspa()</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    title<span class="op">=</span><span class="st">'GSK3B'</span>,          <span class="co"># plot title</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    figsize<span class="op">=</span>(<span class="dv">6</span>, <span class="dv">10</span>),        <span class="co"># figure dimensions (width, height)</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    include_zero<span class="op">=</span><span class="va">False</span>,     <span class="co"># whether to include position 0 in heatmap</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<hr>
<p><strong>Normalization Functions (old version)</strong></p>
<p><code>raw2norm(df, PDHK)</code> — Normalizes raw PSPA data following the method from Johnson et al.&nbsp;(Nature 2022). Column-normalizes by sum of 17 randomized amino acids (excluding S, T, C), scales cysteine by its median, and sets S/T values to position medians.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>norm_df <span class="op">=</span> raw2norm(</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>    df<span class="op">=</span>kinase_df,  <span class="co"># DataFrame with position as index, amino acids as columns</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    PDHK<span class="op">=</span><span class="va">False</span>,    <span class="co"># True for PDHK1/PDHK4 (uses 16 amino acids, also excludes Y)</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><code>get_one_kinase(df, kinase, normalize, drop_s)</code> — Extracts and reshapes data for a single kinase from the stacked PSPA DataFrame, optionally applying normalization.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>kinase_data <span class="op">=</span> get_one_kinase(</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>    df<span class="op">=</span>data,           <span class="co"># stacked DataFrame (kinase as index, position+aa as columns)</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    kinase<span class="op">=</span><span class="st">'PDHK1'</span>,    <span class="co"># kinase name to extract</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    normalize<span class="op">=</span><span class="va">False</span>,   <span class="co"># apply Johnson et al. normalization</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>    drop_s<span class="op">=</span><span class="va">True</span>,       <span class="co"># drop 's' row (duplicate of 't' in PSPA)</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><code>get_logo(df, kinase)</code> — End-to-end function that takes raw stacked PSPA data and generates a sequence logo for a specific kinase, including proper S/T ratio calculation at position 0.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>get_logo(</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>    df<span class="op">=</span>raw_pspa_df,  <span class="co"># stacked DataFrame with kinase as index, substrates as columns</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>    kinase<span class="op">=</span><span class="st">'AAK1'</span>,   <span class="co"># kinase name to visualize</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="setup" class="level2">
<h2 class="anchored" data-anchor-id="setup">Setup</h2>
</section>
<section id="plot" class="level2">
<h2 class="anchored" data-anchor-id="plot">Plot</h2>
<hr>
<section id="preprocess_pspa" class="level3">
<h3 class="anchored" data-anchor-id="preprocess_pspa">preprocess_pspa</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> preprocess_pspa(</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>    pssm</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>):</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Drop row s as it’s a duplicate of t; rename t to pS/pT; calculate np.log2(pssm/pssm.median())</em></p>
<div id="5d5ac217-4043-408b-b336-cfe4fc506446" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>pspa<span class="op">=</span>Data.get_pspa()</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>row <span class="op">=</span> pspa.loc[<span class="st">'GSK3B'</span>]</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>pssm <span class="op">=</span> recover_pssm(row.dropna())</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>pssm <span class="op">=</span> preprocess_pspa(pssm)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>pssm</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">Position</th>
<th data-quarto-table-cell-role="th">-5</th>
<th data-quarto-table-cell-role="th">-4</th>
<th data-quarto-table-cell-role="th">-3</th>
<th data-quarto-table-cell-role="th">-2</th>
<th data-quarto-table-cell-role="th">-1</th>
<th data-quarto-table-cell-role="th">0</th>
<th data-quarto-table-cell-role="th">1</th>
<th data-quarto-table-cell-role="th">2</th>
<th data-quarto-table-cell-role="th">3</th>
<th data-quarto-table-cell-role="th">4</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">aa</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">P</th>
<td>0.128793</td>
<td>0.103768</td>
<td>0.327105</td>
<td>0.377614</td>
<td>0.200697</td>
<td>0.000000</td>
<td>0.857330</td>
<td>-0.156606</td>
<td>-0.022523</td>
<td>0.020985</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">G</th>
<td>0.267518</td>
<td>0.232140</td>
<td>0.709128</td>
<td>0.215152</td>
<td>0.186051</td>
<td>0.000000</td>
<td>-0.070893</td>
<td>-0.132404</td>
<td>-0.032647</td>
<td>-0.138561</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">A</th>
<td>0.198756</td>
<td>0.026650</td>
<td>0.861131</td>
<td>0.210679</td>
<td>0.040110</td>
<td>0.000000</td>
<td>-0.073537</td>
<td>0.007280</td>
<td>-0.146221</td>
<td>-0.092757</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">C</th>
<td>0.281692</td>
<td>0.223789</td>
<td>0.041703</td>
<td>0.297615</td>
<td>-0.036342</td>
<td>0.000000</td>
<td>0.018745</td>
<td>0.043140</td>
<td>0.002481</td>
<td>0.013151</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">S</th>
<td>0.000000</td>
<td>-0.019696</td>
<td>-0.001366</td>
<td>-0.016950</td>
<td>0.000000</td>
<td>5.167707</td>
<td>-0.001258</td>
<td>0.000000</td>
<td>-0.002485</td>
<td>-0.013272</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">T</th>
<td>0.000000</td>
<td>-0.019696</td>
<td>-0.001366</td>
<td>-0.016950</td>
<td>0.000000</td>
<td>4.454046</td>
<td>-0.001258</td>
<td>0.000000</td>
<td>-0.002485</td>
<td>-0.013272</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">V</th>
<td>0.052905</td>
<td>-0.029645</td>
<td>0.009523</td>
<td>0.047180</td>
<td>-0.161873</td>
<td>0.000000</td>
<td>0.067572</td>
<td>-0.151192</td>
<td>-0.154488</td>
<td>-0.213174</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">I</th>
<td>0.000000</td>
<td>-0.135793</td>
<td>-0.045760</td>
<td>0.094081</td>
<td>-0.205347</td>
<td>0.000000</td>
<td>0.065169</td>
<td>-0.034469</td>
<td>-0.089637</td>
<td>-0.222392</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">L</th>
<td>0.041050</td>
<td>-0.054824</td>
<td>-0.194011</td>
<td>-0.100647</td>
<td>-0.161873</td>
<td>0.000000</td>
<td>-0.013903</td>
<td>-0.019595</td>
<td>-0.097586</td>
<td>-0.129861</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">M</th>
<td>-0.049839</td>
<td>-0.090829</td>
<td>-0.012337</td>
<td>-0.100647</td>
<td>0.076869</td>
<td>0.000000</td>
<td>-0.001258</td>
<td>-0.027013</td>
<td>0.100694</td>
<td>-0.026668</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">F</th>
<td>0.071672</td>
<td>-0.127756</td>
<td>-0.144893</td>
<td>-0.145823</td>
<td>-0.247330</td>
<td>0.000000</td>
<td>-0.049908</td>
<td>0.064236</td>
<td>0.098376</td>
<td>-0.092757</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">Y</th>
<td>-0.007367</td>
<td>0.021841</td>
<td>-0.051406</td>
<td>-0.160236</td>
<td>-0.140617</td>
<td>0.000000</td>
<td>-0.192106</td>
<td>0.000000</td>
<td>0.079700</td>
<td>0.114257</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">W</th>
<td>0.108489</td>
<td>0.059871</td>
<td>0.023019</td>
<td>0.016753</td>
<td>-0.091356</td>
<td>0.000000</td>
<td>0.001257</td>
<td>0.007280</td>
<td>0.012363</td>
<td>0.148034</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">H</th>
<td>0.024288</td>
<td>-0.019696</td>
<td>-0.026169</td>
<td>-0.095098</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.016259</td>
<td>0.199623</td>
<td>0.161902</td>
<td>0.089637</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">K</th>
<td>-0.032199</td>
<td>0.144157</td>
<td>0.057528</td>
<td>0.237314</td>
<td>0.094903</td>
<td>0.000000</td>
<td>0.067572</td>
<td>0.016930</td>
<td>0.036774</td>
<td>0.082169</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">R</th>
<td>-0.042252</td>
<td>0.080831</td>
<td>0.598097</td>
<td>0.647946</td>
<td>0.049388</td>
<td>0.000000</td>
<td>0.050669</td>
<td>0.054898</td>
<td>0.084392</td>
<td>-0.129861</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">Q</th>
<td>-0.163205</td>
<td>-0.090829</td>
<td>-0.015093</td>
<td>-0.022206</td>
<td>0.083658</td>
<td>0.000000</td>
<td>0.040921</td>
<td>0.112292</td>
<td>-0.002485</td>
<td>-0.013272</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">N</th>
<td>-0.227709</td>
<td>0.019431</td>
<td>0.001364</td>
<td>-0.016950</td>
<td>0.014284</td>
<td>0.000000</td>
<td>-0.018991</td>
<td>0.156618</td>
<td>-0.007469</td>
<td>0.097067</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">D</th>
<td>-0.271386</td>
<td>-0.049753</td>
<td>-0.001366</td>
<td>-0.040753</td>
<td>-0.210875</td>
<td>0.000000</td>
<td>-0.238837</td>
<td>-0.234608</td>
<td>-0.116305</td>
<td>0.015767</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">E</th>
<td>-0.271386</td>
<td>-0.238787</td>
<td>-0.054238</td>
<td>-0.075844</td>
<td>-0.261600</td>
<td>0.000000</td>
<td>-0.253757</td>
<td>-0.137747</td>
<td>0.315457</td>
<td>1.379934</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">pS/pT</th>
<td>-0.112011</td>
<td>0.206940</td>
<td>0.355120</td>
<td>0.106174</td>
<td>-0.114481</td>
<td>0.000000</td>
<td>0.055519</td>
<td>-0.004874</td>
<td>1.836501</td>
<td>6.163857</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">pY</th>
<td>-0.106727</td>
<td>0.196308</td>
<td>0.158811</td>
<td>0.259140</td>
<td>0.699151</td>
<td>0.000000</td>
<td>0.222213</td>
<td>0.047855</td>
<td>1.191558</td>
<td>1.496894</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<hr>
</section>
<section id="plot_logo_pspa" class="level3">
<h3 class="anchored" data-anchor-id="plot_logo_pspa">plot_logo_pspa</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_logo_pspa(</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>    row, title:<span class="bu">str</span><span class="op">=</span><span class="st">'Motif'</span>, figsize:<span class="bu">tuple</span><span class="op">=</span>(<span class="dv">5</span>, <span class="dv">2</span>)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>):</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div id="38a1a3cb-88b1-43d3-bc78-24c9920b5136" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>plot_logo_pspa(pspa.loc[<span class="st">'GSK3B'</span>],title<span class="op">=</span><span class="st">'GSK3B'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="02d_pssm_pspa_files/figure-html/cell-5-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
<section id="plot_logo_heatmap_pspa" class="level3">
<h3 class="anchored" data-anchor-id="plot_logo_heatmap_pspa">plot_logo_heatmap_pspa</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_logo_heatmap_pspa(</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>    row, <span class="co"># row of Data.get_pspa()</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>    title:<span class="bu">str</span><span class="op">=</span><span class="st">'Motif'</span>, figsize:<span class="bu">tuple</span><span class="op">=</span>(<span class="dv">6</span>, <span class="dv">10</span>), include_zero:<span class="bu">bool</span><span class="op">=</span><span class="va">False</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>):</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Plot logo and heatmap vertically</em></p>
<div id="0a9abfde-2b65-4bb2-ab35-31c5abc87072" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>plot_logo_heatmap_pspa(pspa.loc[<span class="st">'GSK3B'</span>],title<span class="op">=</span><span class="st">'GSK3B'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="02d_pssm_pspa_files/figure-html/cell-7-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="others-old-version" class="level2">
<h2 class="anchored" data-anchor-id="others-old-version">Others (old version)</h2>
<hr>
<section id="raw2norm" class="level3">
<h3 class="anchored" data-anchor-id="raw2norm">raw2norm</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> raw2norm(</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    df:DataFrame, <span class="co"># single kinase's df has position as index, and single amino acid as columns</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    PDHK:<span class="bu">bool</span><span class="op">=</span><span class="va">False</span>, <span class="co"># whether this kinase belongs to PDHK family</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>):</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Normalize single ST kinase data</em></p>
<p>This function implement the normalization method from <a href="https://www.nature.com/articles/s41586-022-05575-3#Sec6">Johnson et al.&nbsp;Nature: An atlas of substrate specificities for the human serine/threonine kinome</a></p>
<p>Specifically, &gt; - matrices were column-normalized at all positions by the sum of the 17 randomized amino acids (excluding serine, threonine and cysteine), to yield PSSMs. &gt;- PDHK1 and PDHK4 were normalized to the 16 randomized amino acids (excluding serine, threonine, cysteine and additionally tyrosine) &gt;- The cysteine row was scaled by its median to be 1/17 (1/16 for PDHK1 and PDHK4). &gt;- The serine and threonine values in each position were set to be the median of that position. &gt;- The S0/T0 ratio was determined by summing the values of S and T rows in the matrix (SS and ST, respectively), accounting for the different S vs.&nbsp;T composition of the central (1:1) and peripheral (only S or only T) positions (Sctrl and Tctrl, respectively), and then normalizing to the higher value among the two (S0 and T0, respectively, Supplementary Note 1)</p>
<p>This function is usually implemented with the below function, with <code>normalize</code> being a bool argument.</p>
<hr>
</section>
<section id="get_one_kinase" class="level3">
<h3 class="anchored" data-anchor-id="get_one_kinase">get_one_kinase</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_one_kinase(</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>    df:DataFrame, <span class="co"># stacked dataframe (paper's raw data)</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    kinase:<span class="bu">str</span>, <span class="co"># a specific kinase</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    normalize:<span class="bu">bool</span><span class="op">=</span><span class="va">False</span>, <span class="co"># normalize according to the paper; special for PDHK1/4</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>    drop_s:<span class="bu">bool</span><span class="op">=</span><span class="va">True</span>, <span class="co"># drop s as s is a duplicates of t in PSPA</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>):</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Obtain a specific kinase data from stacked dataframe</em></p>
<p>Retreive a single kinase data from PSPA data that has an format of kinase as index and position+amino acid as column.</p>
<div id="3a78e9b2-13fa-4a42-848d-eaefb1ef4dd0" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>data <span class="op">=</span> Data.get_pspa_st()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="f0382a84-b5b0-4354-9726-98d05cfd6acf" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>get_one_kinase(data,<span class="st">'PDHK1'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">aa</th>
<th data-quarto-table-cell-role="th">A</th>
<th data-quarto-table-cell-role="th">C</th>
<th data-quarto-table-cell-role="th">D</th>
<th data-quarto-table-cell-role="th">E</th>
<th data-quarto-table-cell-role="th">F</th>
<th data-quarto-table-cell-role="th">G</th>
<th data-quarto-table-cell-role="th">H</th>
<th data-quarto-table-cell-role="th">I</th>
<th data-quarto-table-cell-role="th">K</th>
<th data-quarto-table-cell-role="th">L</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">P</th>
<th data-quarto-table-cell-role="th">Q</th>
<th data-quarto-table-cell-role="th">R</th>
<th data-quarto-table-cell-role="th">S</th>
<th data-quarto-table-cell-role="th">T</th>
<th data-quarto-table-cell-role="th">V</th>
<th data-quarto-table-cell-role="th">W</th>
<th data-quarto-table-cell-role="th">Y</th>
<th data-quarto-table-cell-role="th">t</th>
<th data-quarto-table-cell-role="th">y</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">position</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">-5</th>
<td>0.0594</td>
<td>0.0625</td>
<td>0.0589</td>
<td>0.0550</td>
<td>0.0775</td>
<td>0.0697</td>
<td>0.0687</td>
<td>0.0590</td>
<td>0.0515</td>
<td>0.0657</td>
<td>...</td>
<td>0.0451</td>
<td>0.0424</td>
<td>0.0594</td>
<td>0.0594</td>
<td>0.0594</td>
<td>0.0573</td>
<td>0.1001</td>
<td>0.0775</td>
<td>0.0583</td>
<td>0.0658</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">-4</th>
<td>0.0618</td>
<td>0.0621</td>
<td>0.0550</td>
<td>0.0511</td>
<td>0.0739</td>
<td>0.0715</td>
<td>0.0598</td>
<td>0.0601</td>
<td>0.0520</td>
<td>0.0614</td>
<td>...</td>
<td>0.0637</td>
<td>0.0552</td>
<td>0.0617</td>
<td>0.0608</td>
<td>0.0608</td>
<td>0.0519</td>
<td>0.0916</td>
<td>0.0739</td>
<td>0.0528</td>
<td>0.0752</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">-3</th>
<td>0.0608</td>
<td>0.0576</td>
<td>0.0499</td>
<td>0.0423</td>
<td>0.0803</td>
<td>0.0580</td>
<td>0.0674</td>
<td>0.0687</td>
<td>0.0481</td>
<td>0.0667</td>
<td>...</td>
<td>0.0570</td>
<td>0.0532</td>
<td>0.0532</td>
<td>0.0584</td>
<td>0.0584</td>
<td>0.0588</td>
<td>0.1113</td>
<td>0.0803</td>
<td>0.0416</td>
<td>0.0553</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">-2</th>
<td>0.0587</td>
<td>0.0655</td>
<td>0.0470</td>
<td>0.0437</td>
<td>0.0790</td>
<td>0.0890</td>
<td>0.0787</td>
<td>0.0533</td>
<td>0.0440</td>
<td>0.0637</td>
<td>...</td>
<td>0.0500</td>
<td>0.0543</td>
<td>0.0616</td>
<td>0.0565</td>
<td>0.0565</td>
<td>0.0519</td>
<td>0.1082</td>
<td>0.0790</td>
<td>0.0327</td>
<td>0.0557</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">-1</th>
<td>0.0782</td>
<td>0.1009</td>
<td>0.0989</td>
<td>0.0426</td>
<td>0.0650</td>
<td>0.0695</td>
<td>0.0782</td>
<td>0.0496</td>
<td>0.0409</td>
<td>0.0578</td>
<td>...</td>
<td>0.0540</td>
<td>0.0500</td>
<td>0.0469</td>
<td>0.0594</td>
<td>0.0594</td>
<td>0.0514</td>
<td>0.0756</td>
<td>0.0650</td>
<td>0.0358</td>
<td>0.0433</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">0</th>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>...</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>1.0000</td>
<td>0.4886</td>
<td>NaN</td>
<td>NaN</td>
<td>0.0000</td>
<td>0.4886</td>
<td>0.0000</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">1</th>
<td>0.0400</td>
<td>0.0562</td>
<td>0.0394</td>
<td>0.0355</td>
<td>0.0735</td>
<td>0.0400</td>
<td>0.0502</td>
<td>0.1288</td>
<td>0.0390</td>
<td>0.1439</td>
<td>...</td>
<td>0.0379</td>
<td>0.0455</td>
<td>0.0455</td>
<td>0.0455</td>
<td>0.0455</td>
<td>0.0797</td>
<td>0.0784</td>
<td>0.0735</td>
<td>0.0336</td>
<td>0.0452</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">2</th>
<td>0.0496</td>
<td>0.0783</td>
<td>0.0643</td>
<td>0.0555</td>
<td>0.0720</td>
<td>0.1067</td>
<td>0.0684</td>
<td>0.0480</td>
<td>0.0505</td>
<td>0.0555</td>
<td>...</td>
<td>0.0564</td>
<td>0.0653</td>
<td>0.0695</td>
<td>0.0601</td>
<td>0.0601</td>
<td>0.0508</td>
<td>0.0672</td>
<td>0.0720</td>
<td>0.0414</td>
<td>0.0594</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">3</th>
<td>0.0486</td>
<td>0.0609</td>
<td>0.0938</td>
<td>0.0684</td>
<td>0.1024</td>
<td>0.0676</td>
<td>0.0544</td>
<td>0.0583</td>
<td>0.0388</td>
<td>0.0552</td>
<td>...</td>
<td>0.0686</td>
<td>0.0502</td>
<td>0.0561</td>
<td>0.0588</td>
<td>0.0588</td>
<td>0.0593</td>
<td>0.0641</td>
<td>0.1024</td>
<td>0.0539</td>
<td>0.0431</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">4</th>
<td>0.0565</td>
<td>0.0749</td>
<td>0.0631</td>
<td>0.0535</td>
<td>0.0732</td>
<td>0.0655</td>
<td>0.0664</td>
<td>0.0625</td>
<td>0.0496</td>
<td>0.0552</td>
<td>...</td>
<td>0.0677</td>
<td>0.0553</td>
<td>0.0604</td>
<td>0.0626</td>
<td>0.0626</td>
<td>0.0579</td>
<td>0.0864</td>
<td>0.0732</td>
<td>0.0548</td>
<td>0.0575</td>
</tr>
</tbody>
</table>

<p>10 rows × 22 columns</p>
</div>
</div>
</div>
<hr>
</section>
<section id="get_logo" class="level3">
<h3 class="anchored" data-anchor-id="get_logo">get_logo</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_logo(</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>    df:DataFrame, <span class="co"># stacked Dataframe with kinase as index, substrates as columns</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>    kinase:<span class="bu">str</span>, <span class="co"># a specific kinase name in index</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>):</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><em>Given stacked df (index as kinase, columns as substrates), get a specific kinase’s logo</em></p>
<p>This function is to replicate the motif logo from <a href="https://www.nature.com/articles/s41586-022-05575-3">Johnson et al.&nbsp;Nature: An atlas of substrate specificities for the human serine/threonine kinome</a>. Given raw PSPA data, it can output a motif logo.</p>
<div id="b2634aae-764f-455a-8b3b-b2308f314a11" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load raw PSPA data</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="co"># df = pd.read_csv('https://github.com/sky1ove/katlas_raw/raw/refs/heads/main/nbs/raw/pspa_st_raw.csv').set_index('kinase')</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="co"># df.head()</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="co"># get_logo(df, 'AAK1')</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/sky1ove\.github\.io\/katlas");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/sky1ove/katlas/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>